rhumba/backends/zk.py

```diff
@@ -0,0 +1,453 @@
+            if beat:
+                if time.time() - float(beat) < 10:
+                    snames.append(server)
+
+        defer.returnValue(snames)
+
+    @defer.inlineCallbacks
+    def clusterQueues(self):
+        """ Return a dict of queues in cluster and servers running them
+        """
+        servers = yield self.getClusterServers()
+
+        queues = {}
+
+        for sname in servers:
+            qs = yield self._get_key('/server/%s/queues' % sname)
+            uuid = yield self._get_key('/server/%s/uuid' % sname)
+       
+            qs = json.loads(qs)
+
+            for q in qs:
+                if q not in queues:
+                    queues[q] = []
+
+                queues[q].append({'host': sname, 'uuid': uuid})
+
+        defer.returnValue(queues)
+
+    @defer.inlineCallbacks
+    def clusterStatus(self):
+        """
+        Returns a dict of cluster nodes and their status information
+        """
+        servers = yield self.getClusterServers()
+
+        d = {
+            'workers': {},
+            'crons': {},
+            'queues': {}
+        }
+
+        now = time.time()
+
+        reverse_map = {}
+
+        for sname in servers:
+            last = yield self._get_key('/server/%s/heartbeat' % sname)
+            status = yield self._get_key('/server/%s/status' % sname)
+            uuid = yield self._get_key('/server/%s/uuid' % sname)
+
+            reverse_map[uuid] = sname
+
+            if not last:
+                last = 0
+
+            last = float(last)
+
+            if (status == 'ready') and (now - last > 5):
+                status = 'offline'
+
+            if not sname in d['workers']:
+                d['workers'][sname] = []
+
+            d['workers'][sname].append({
+                'lastseen': last,
+                'status': status,
+                'id': uuid
+            })
+
+        # Crons
+        crons = yield self.keys('/crons')
+
+        for queue in crons:
+            if queue not in d['crons']:
+                d['crons'][queue] = {'methods': {}}
+
+            methods = yield self.keys('/crons/%s' % queue)
+
+            for method in methods:
+                last = yield self._get_key('/crons/%s/%s' % (queue, method))
+                if last:
+                    d['crons'][queue]['methods'][method] = float(last)
+            
+            uid = yield self._get_key('/croner/%s' % queue)
+            if uid:
+                d['crons'][queue]['master'] = '%s:%s' % (uid, reverse_map[uid])
+
+        # Queues
+        queue_keys = yield self.keys('/qstats')
+
+        for qname in queue_keys:
+            if qname not in d['queues']:
+                qlen = yield self.queueSize(qname)
+
+                stats = yield self.getQueueMessageStats(qname)
+
+                d['queues'][qname] = {
+                    'waiting': qlen,
+                    'messages': stats
+                }
+
+        defer.returnValue(d)
```

Comment:
Can we break up this gigantic method a bit? Maybe factor out some of the loop bodies into separate functions.